src/appleseed/renderer/modeling/bsdf/alsurfacelayerbrdf.cpp

```diff
@@ -0,0 +1,482 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "alsurfacelayerbrdf.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/lighting/scatteringmode.h"
+#include "renderer/kernel/shading/closures.h"
+#include "renderer/kernel/shading/shadingcontext.h"
+#include "renderer/modeling/bsdf/bsdf.h"
+#include "renderer/modeling/bsdf/bsdfwrapper.h"
+#include "renderer/modeling/bsdf/fresnel.h"
+#include "renderer/modeling/bsdf/microfacethelper.h"
+#include "renderer/modeling/bsdf/specularhelper.h"
+#include "renderer/utility/messagecontext.h"
+#include "renderer/utility/paramarray.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/basis.h"
+#include "foundation/math/microfacet.h"
+#include "foundation/math/minmax.h"
+#include "foundation/math/sampling/mappings.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+#include "foundation/utility/makevector.h"
+#include "foundation/utility/otherwise.h"
+
+// Standard headers.
+#include <algorithm>
+#include <cmath>
+#include <string>
+
+// Forward declarations.
+namespace foundation    { class IAbortSwitch; }
+namespace renderer      { class Assembly; }
+namespace renderer      { class Project; }
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // AlSurfaceLayer BRDF.
+    //
+    // References:
+    //
+    //  https://bitbucket.org/anderslanglands/alshaders/wiki/Home
+    //  http://blog.selfshadow.com/publications/s2014-shading-course/langlands/s2014_pbs_alshaders_notes.pdf
+    //
+
+    const char* Model = "al_surface_layer_brdf";
+
+    class AlSurfaceLayerBRDFImpl
+      : public BSDF
+    {
+      public:
+        AlSurfaceLayerBRDFImpl(
+            const char*             name,
+            const ParamArray&       params)
+          : BSDF(name, AllBSDFTypes, ScatteringMode::All, params)
+        {
+        }
+
+        virtual void release() APPLESEED_OVERRIDE
+        {
+            delete this;
+        }
+
+        virtual const char* get_model() const APPLESEED_OVERRIDE
+        {
+            return Model;
+        }
+
+        virtual size_t compute_input_data_size(
+            const Assembly&         assembly) const APPLESEED_OVERRIDE
+        {
+            return align(sizeof(InputValues), 16);
+        }
+
+        APPLESEED_FORCE_INLINE virtual void prepare_inputs(
+            const ShadingContext&   shading_context,
+            const ShadingPoint&     shading_point,
+            void*                   data) const APPLESEED_OVERRIDE
+        {
+            InputValues* values = reinterpret_cast<InputValues*>(data);
+
+            new (&values->m_precomputed) InputValues::Precomputed();
+
+            const float roughness = square(values->m_roughness);
```

Comment:
Rename `roughness` to `alpha`? Otherwise it's quite confusing, since everywhere else roughness is the square root of alpha...